tests/generation/test_utils.py

```diff
@@ -1236,6 +1236,69 @@ def test_prompt_lookup_decoding_matches_greedy_search(self):
                 self._check_outputs(output, input_ids, model.config, use_cache=True)
 
+    def test_dola_decoding_sample(self):
+        # TODO (joao): investigate skips, try to reduce incompatibilities
+        for model_class in self.all_generative_model_classes:
+            if any(model_name in model_class.__name__.lower() for model_name in ["reformer"]):
+                self.skipTest("Skip Reformer as the lm_head input size is 2 * hidden size, adopted from Rev Nets.")
+            if any(
+                model_name in model_class.__name__.lower()
+                for model_name in [
+                    "wav2vec",
+                    "clvp",
+                    "bark",
+                ]
+            ):
+                self.skipTest("Skip speech models")
+            if any(
+                model_name in model_class.__name__.lower()
+                for model_name in [
+                    "marian",
+                    "mbart",
+                    "pegasus",
+                ]
+            ):
+                self.skipTest("DoLa is not supported for the models without returning layerwise hidden states")
+
+            # enable cache if the model is not openai-gpt, xlnet, cpm, or xlm
+            config, input_ids, attention_mask, _ = self._get_input_ids_and_config(batch_size=1)
+            config.use_cache = True
+            if any(
+                model_name in model_class.__name__.lower()
+                for model_name in [
+                    "openaigpt",
+                    "xlnet",
+                    "cpm",
+                    "xlm",
+                ]
+            ):
+                config.use_cache = False
+
+            # Encoder-decoder models are not supported
+            if config.is_encoder_decoder:
+                self.skipTest("DoLa is not supported for encoder-decoder models")
+            config.is_decoder = True
+            model = model_class(config).to(torch_device).eval()
+
+            # Sets dola generation arguments such that:
+            # a) no EOS is generated, to ensure generation doesn't break early
+            # b) there are at least two forward passes in the main model, to ensure the input preparation of
+            #    the main model is correct
+            generation_kwargs = {
+                "eos_token_id": -1,  # see a)
+                "max_new_tokens": 4,  # see b)
+                "num_beams": 1,
+                "do_sample": True,
+                "output_scores": True,
+                "output_logits": True,
+                "output_hidden_states": True,
+                "output_attentions": True,
+                "return_dict_in_generate": True,
+            }
+            generation_kwargs.update({"dola_layers": "low"})
+            output_dola = model.generate(input_ids, attention_mask=attention_mask, **generation_kwargs)
+            self._check_outputs(output_dola, input_ids, model.config, use_cache=config.use_cache)
```

Comment:
We should be able to do a test which does a single forward pass and checks that the expected logits are selected i.e. the dola method should be decoupled from generate itself and we test passing logits to the dola method and then the logit outputs. I believe this is a more general issue with the generation testing however. 

Specifically, this test doesn't really convince me that the implementation is correct (not do the integration tests, unless they've been generated from the official dola implementation), but that they functionally work

Reply:
For the correctness of DoLa. I am the first author of DoLa paper and I have kept tracking whether the new code in this PR can reproduce the old numbers in my paper.

<img width="633" alt="image" src="https://github.com/huggingface/transformers/assets/26344602/1081733d-4889-4f8c-908c-1d8de30edb4b">

The left-hand side is the new numbers I tested using the current version of code. 
The right-hand side is the screenshot of my paper, where the numbers are from the [official implementation](https://github.com/voidism/DoLa) and the experiments I did last year. 

The original implementation was based on `v4.28.1`. The numbers changed a little bit (also for the greedy decoding baseline), which I think it's because of the version changes as well as the different machines and gpus I used. But the same level of improvement can be achieved by the new code in this PR, e.g. ~4% on StrQA with llama-7b. 

I can also provide more tests to validate the consistency between this PR and my official dola implementation if you think it's needed!

Reply:
I have checked that my latest commit today (based on `v4.41.0`) can also reproduce the scores here!